src/worker.rs

```diff
@@ -0,0 +1,409 @@
+        self
+    }
+
+    /// shutdown is a builder method that allows to modify the Shutdown settings
+    /// for the spawned worker
+    pub fn shutdown(mut self, shutdown: Shutdown) -> Self {
+        self.shutdown = shutdown;
+        self
+    }
+
+    pub fn start_timeout(mut self, timeout: Duration) -> Self {
+        self.start_timeout = timeout;
+        self
+    }
+
+    pub fn termination_timeout(mut self, timeout: Duration) -> Self {
+        self.termination_timeout = timeout;
+        self
+    }
+
+    /// start spawns a new routine that executes the start function of this
+    /// Spec.
+    ///
+    /// ### Blocking on start function
+    ///
+    /// This function will block and wait until a notification from the
+    /// start function (executing on a newly spawned routine) is received; this
+    /// notification indicates if the start was successful or not.
+    ///
+    /// This blocking is necessary to ensure that supervised worker trees get
+    /// spawned in the specified/expected order.
+    pub async fn start(
+        mut self,
+        parent_ctx: &Context,
+        parent_name: &str,
+    ) -> anyhow::Result<Worker> {
+        let runtime_name = format!("{}/{}", parent_name, self.name);
+        let created_at = Utc::now();
+
+        // ON_START setup
+        let (started_tx, started_rx) = oneshot::channel::<Result<(), anyhow::Error>>();
+        let start_notifier = StartNotifier::from_oneshot(started_tx);
+
+        // CANCEL setup
+        let (ctx, termination_handle) = Context::with_cancel(parent_ctx);
+        let (kill_handle, kill_registration) = AbortHandle::new_pair();
+        let future = (*self.routine)(ctx, start_notifier);
+        let routine = Abortable::new(future, kill_registration);
+
+        // SPAWN -- actually spawn concurrent worker future
+        let join_handle = task::spawn(routine);
+
+        // ON_START blocking -- block before moving to the next start
+        let err = time::timeout(self.start_timeout, started_rx).await;
+
+        match err {
+            // On worker initialization failed, we need to signal
+            // this to starter
+            Err(start_timeout_err) => Err(anyhow::Error::new(start_timeout_err)),
+
+            Ok(Err(start_notify_not_invoked_err)) => {
+                Err(anyhow::Error::new(start_notify_not_invoked_err))
+            }
+
+            Ok(Ok(Err(routine_err))) => Err(routine_err),
+
+            // Happy path
+            Ok(Ok(Ok(_))) => Ok(Worker {
+                spec: self,
+                runtime_name,
+                created_at,
+                join_handle,
+                termination_handle,
+                kill_handle,
+            }),
+        }
+    }
+}
+
+impl Worker {
+    /// terminate tries to stop gracefuly a worker routine. In the scenario that
+    /// the routine doesn't stop after a timeout, it is brutally killed.
+    pub async fn terminate(self) -> (Spec, Option<Arc<anyhow::Error>>) {
```

Comment:
Why is the error wrapped in an `Arc` I presume that means it's going to be shared somehow?

Reply:
I was hoping you asked for that, yes... before getting to the supervision bits, I'm planning to implement an event system to add observability to the supervision tree as a whole. The errors need to be shared to the events that get emitted as well as being used for error handling.

Is this the right way to go?

Reply:
Yeah, I think it's at least the right place to start.